### PR TITLE
builtins: implement ST_Force3D, ST_Force3DM, ST_Force3DZ, ST_Force4D

### LINKED ISSUE #60880: geo/geomfn: implement ST_Force3DM({geometry})
Implement `ST_Force3DM` on arguments {geometry}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_Force3DM.html).

_Observers: Please react to this issue if you need this functionality._

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

**The following additional guidance has been issued on implementing this function:**

use ForceLayout

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2021-02-21T20:42:26Z. Changes to titles, body and labels may be overwritten.</sub>

### LINKED ISSUE #60881: geo/geomfn: implement ST_Force3DZ({geometry})
Implement `ST_Force3DZ` on arguments {geometry}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_Force3DZ.html).

_Observers: Please react to this issue if you need this functionality._

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

**The following additional guidance has been issued on implementing this function:**

use ForceLayout

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2021-02-21T20:42:26Z. Changes to titles, body and labels may be overwritten.</sub>

### LINKED ISSUE #60882: geo/geomfn: implement ST_Force4D({geometry})
Implement `ST_Force4D` on arguments {geometry}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_Force4D.html).

_Observers: Please react to this issue if you need this functionality._

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

**The following additional guidance has been issued on implementing this function:**

use ForceLayout

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2021-02-21T20:42:26Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by andyyang890**
This patch implements the geometry builtins `ST_Force3D`,
`ST_Force3DM`, `ST_Force3DZ`, and `ST_Force4D`.

Resolves #60880.
Resolves #60881.
Resolves #60882.

Release justification: low-risk update to new functionality
Release note (sql change): Geometry builtins for forcing geometries
into non-2D layouts are now available.

### CODE REVIEW

**Review by otan — APPROVED** [review]

File: pkg/geo/geomfn/force_layout.go [hunk 1]
```diff
@@ -17,29 +17,60 @@ import (
...
+	retGeomT, err := forceLayout(geomT, layout, defaultZ, defaultM)
 	if err != nil {
 		return geo.Geometry{}, err
 	}
 	return geo.MakeGeometryFromGeomT(retGeomT)
 }
 
-// forceLayout forces a geom.T into the given layout.
-func forceLayout(t geom.T, layout geom.Layout) (geom.T, error) {
+// forceLayout forces a geom.T into the given layout and uses the specified default Z and M values as needed.
```
Comment (rafiss) [inline pkg/geo/geomfn/force_layout.go:61]: small nit: we wrap comments to 80 characters (and code to 100 characters). i use the wrap to column plugin for comments https://plugins.jetbrains.com/plugin/7234-wrap-to-column

**Review by rafiss — APPROVED** [review]
lgtm! left a comment on comment formatting

File: pkg/geo/geomfn/force_layout.go (hunk 1 above)
Comment (andyyang890) [reply pkg/geo/geomfn/force_layout.go]: thanks for the tip :) I actually have this plugin installed already, not sure why it didn't wrap the comment automatically. i've manually broken the comment up into two lines.

**Comment by andyyang890** [comment]
bors r=otan,rafiss

**Pull request merged**